File: app/src/main/java/fiix/challenge/fiixexercise/javasample/MainFragment.java
package fiix.challenge.fiixexercise.javasample;


import android.arch.lifecycle.Observer;
import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;
import java.util.List;

import fiix.challenge.fiixexercise.R;

public class MainFragment extends Fragment {
    private MainViewModel mViewModel;
    private RecyclerView recyclerViewView;
    //private Context context;
    private RecyclerAdapter recyclerAdapter;
    private List<TriviaQuestion2> qaList;

    public static MainFragment newInstance() {
        return new MainFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_main, container, false);
        recyclerViewView = v.findViewById(R.id.my_recycler_view);
        qaList = new ArrayList<>();
        return v;
    }


    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mViewModel = ViewModelProviders.of(getActivity()).get(MainViewModel.class);
        mViewModel.getQa().observe(getActivity(), new Observer<List<TriviaQuestion2>>() {
            @Override
            public void onChanged(@Nullable List<TriviaQuestion2> questionAndAnswers) {
                qaList = questionAndAnswers;
                RecyclerView.LayoutManager layoutManager = new LinearLayoutManager(getActivity());
                recyclerViewView.setLayoutManager(layoutManager);
                recyclerAdapter = new RecyclerAdapter(qaList);
                recyclerViewView.setHasFixedSize(true);
                recyclerViewView.setAdapter(recyclerAdapter);
                recyclerViewView.setVisibility(View.VISIBLE);

            }
        });


    }

}

File: app/src/main/java/fiix/challenge/fiixexercise/kotlinsample/ui/MainFragment.kt
package fiix.challenge.fiixexercise.kotlinsample.ui

import android.content.Context
import androidx.lifecycle.ViewModelProvider
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import fiix.challenge.fiixexercise.R
import fiix.challenge.fiixexercise.databinding.MainFragmentBinding
import fiix.challenge.fiixexercise.dp.DataProcessor
import fiix.challenge.fiixexercise.kotlinsample.LocalDataSource
import fiix.challenge.fiixexercise.kotlinsample.MainActivity
import fiix.challenge.fiixexercise.kotlinsample.MockRepo
import fiix.challenge.fiixexercise.kotlinsample.TriviaQuestion
import fiix.challenge.fiixexercise.kotlinsample.adapter.QuestionsAdapter
import fiix.challenge.fiixexercise.kotlinsample.db.QuestionsDatabase
import fiix.challenge.fiixexercise.kotlinsample.model.Resource
import fiix.challenge.fiixexercise.kotlinsample.repository.QuestionsRepository
import kotlinx.android.synthetic.main.activity_main.*

class MainFragment : Fragment() {

    companion object {
        fun newInstance() = MainFragment()
    }

    private lateinit var binding: MainFragmentBinding
    private lateinit var questionAdapter: QuestionsAdapter
    private lateinit var viewModel: MainViewModel

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = MainFragmentBinding.inflate(layoutInflater)
        return binding.root
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = (activity as MainActivity).viewModel

        setupRecyclerView()

        viewModel.questions.observe(viewLifecycleOwner, { response->
            when(response){
                is Resource.Success -> {
                    toggleProgressBar(true)
                }

                is Resource.Loading -> {
                    toggleProgressBar(false)
                }
            }

        })

        viewModel.getQuestionsInDb().observe(viewLifecycleOwner, {
            questionAdapter.diff.submitList(it)
        })
    }

    private fun toggleProgressBar(hide: Boolean) {
        if (hide)
            binding.pbProgress.visibility = View.INVISIBLE
        else
            binding.pbProgress.visibility = View.VISIBLE

    }

    private fun setupRecyclerView() {
        questionAdapter = QuestionsAdapter()
        questionAdapter.setOnListItemClickListener { question->
            val b = Bundle()
            b.putParcelable("question", question)
            findNavController().navigate(R.id.action_mainFragment_to_editFragment, b)

        }
        questionAdapter.setOnAnswerBtnClickListener { item->
            item.answerShown = true
            viewModel.updateQuestionInDb(item)
        }

        binding.rvQuestions.apply{
            adapter = questionAdapter
            layoutManager = LinearLayoutManager(activity)
        }

    }

}